lib/backends/webgl/utils.ts

```diff
@@ -11,3 +11,53 @@ export function getPackedShape(unpackedShape: ReadonlyArray<number>): ReadonlyAr
   const len = unpackedShape.length;
   return unpackedShape.slice(0, len - 1).concat(unpackedShape[len - 1] / 4);
 }
+
+/**
+ * Generates the function name from an input sampler name.
+ * @param samplerName Name of the sampler.
+ */
+export function generateShaderFuncNameFromInputSamplerName(samplerName: string): string {
+  assert(typeof samplerName !== 'undefined' && samplerName.length !== 0, () => 'empty string found for sampler name');
+  return 'get' + samplerName.charAt(0).toUpperCase() + samplerName.slice(1);
+}
+
+/**
+ * Generates the function name from an input sampler name at output coordinates.
+ * @param samplerName Name of the sampler.
+ */
+export function generateShaderFuncNameFromInputSamplerNameAtOutCoords(samplerName: string): string {
+  assert(typeof samplerName !== 'undefined' && samplerName.length !== 0, () => 'empty string found for sampler name');
+  return 'get' + samplerName.charAt(0).toUpperCase() + samplerName.slice(1) + 'AtOutCoords';
+}
+
+/** Returns a new input shape (a copy) that has a squeezed logical shape. */
+export function squeezeInputShape(inputShape: ReadonlyArray<number>, squeezedShape: number[]): number[] {
+  // Deep copy.
+  let newInputShape: number[] = JSON.parse(JSON.stringify(inputShape));
```

Comment:
> let newInputShape: number[] = JSON.parse(JSON.stringify(inputShape)); [](start = 1, length = 70)

[yulong] consider using slice if JSON is a heavy library. And check other occurance where JSON library is used.

Reply:
this needs to make a deep copy of the input shape while slice returns a shallow copy. The input shape is usually of a small array so perf slow-down is neglectable.